third_party/amd/lib/TritonAMDGPUToLLVM/LoadStoreOpToLLVM.cpp

```diff
@@ -446,6 +398,123 @@ struct DirectToLdsLoadConversionBase : public LoadStoreConversionBase {
+      packedArr = b.insert_val(packedArr, srcElems[i], 0);
+      // mask
+      auto maskElem = maskElems.empty() ? b.true_val() : maskElems[i];
+      packedArr = b.insert_val(packedArr, maskElem, 1);
+      // other
+      if (!otherElems.empty())
+        packedArr = b.insert_val(packedArr, otherElems[i], 2);
+      // swizzleOffset are per vec so we need to duplicate values vec times
+      auto swizzleOffset = swizzledLaneOffsets.empty()
+                               ? b.i32_val(0)
+                               : swizzledLaneOffsets[i / vec];
+      packedArr = b.insert_val(packedArr, swizzleOffset, 3);
+
+      loadVals.push_back(packedArr);
+    }
+    return loadVals;
+  }
+
+  auto unzipLoadValues(RewriterBase &rewriter, Location loc, int startIdx,
+                       ArrayRef<Value> values, Type srcTy, Type otherTy,
+                       bool hasOther, unsigned vec) const {
+    TritonLLVMOpBuilder b(loc, rewriter);
+    auto structElem = values[startIdx];
+    Value offsetElem = b.extract_val(srcTy, structElem, 0);
+    Value maskElem = b.extract_val(i1_ty, structElem, 1);
+    // Gather other elements
+    SmallVector<Value> otherElems;
+    if (hasOther) {
+      for (int i = 0; i < vec; i++) {
+        otherElems.push_back(b.extract_val(otherTy, values[startIdx + i], 2));
+      }
+    }
+
+    Value swizzleLaneOffset = b.extract_val(i32_ty, structElem, 3);
+
+    return std::make_tuple(offsetElem, maskElem, std::move(otherElems),
+                           swizzleLaneOffset);
+  }
+
+  void applySwizzling(RewriterBase &rewriter, Location loc, Value &srcOrOffset,
+                      Value &mask, Value laneId,
+                      Value swizzleLaneOffset) const {
+    TritonLLVMOpBuilder b(loc, rewriter);
+    // laneId + swizzleOffset will always stay inside the warp [0,
+    // threadsPerWarp) because we only swizzle inside a warp
+    Value swizzledLaneId = b.add(laneId, swizzleLaneOffset);
+    // Shuffle based on swizzleLaneId to apply the swizzling
+    srcOrOffset =
+        targetInfo.shuffleIdx(rewriter, loc, srcOrOffset, swizzledLaneId);
+
+    if (mask) {
+      mask = shuffleMask(rewriter, b, loc, targetInfo, swizzledLaneId, mask);
+    }
+  }
+
+  void lowerDirectToLDSLoad(
+      RewriterBase &rewriter, Location loc, RankedTensorType srcTy,
+      MemDescType dstTy, SmallVector<Value> loadVals, Value llDst,
+      Type resElemTy, unsigned vec,
+      std::function<SmallVector<Value>(RewriterBase &, Location,
+                                       ArrayRef<Value>, Value, int, VectorType)>
+          lowerInst) const {
+    TritonLLVMOpBuilder b(loc, rewriter);
+    auto *ctx = rewriter.getContext();
+
+    // Build src to shared layout and remove broadcasted registers
+    auto srcLayout = triton::gpu::toLinearLayout(srcTy);
+    auto removeBroadcastSrc = actionRemoveBroadcastedRegs(srcLayout);
+    srcLayout = removeBroadcastSrc.apply(srcLayout);
+    loadVals = removeBroadcastSrc.apply(loadVals);
```

Comment:
Why don't you handle here the element broadcasting?

Reply:
I am probably missing something obvious here but I am not sure what we should handle here in terms of broadcasting?
We are removing the broadcasting reg bases since they would emit redundant loads to the same LDS addresses?

Reply:
sure, but if you are removing them, then you should:
1. In the store, remove the associated values
2. In the load, duplicate the values after the load before packing

See
https://github.com/triton-lang/triton/blob/1af80393d9f0903404dee85e81f226ee9365f53c/lib/Conversion/TritonGPUToLLVM/Utility.cpp#L713-L727

You must be doing something like this somewhere, but these utils allow you to do it in a rather compact way.

Reply:
yeah @AlexAUT please send follow up PR to refactor as discussed here later :)